socks/socks.go

```diff
@@ -162,30 +159,33 @@ func ParseAddr(s string) Addr {
 }
 
 // Handshake fast-tracks SOCKS initialization to get target address to connect.
+// Read RFC 1928 for request and reply structure and sizes.
 func Handshake(rw io.ReadWriter) (Addr, error) {
-	// Read RFC 1928 section 4 for request and reply structure and sizes
-	buf := make([]byte, MaxReqLen)
-
-	_, err := rw.Read(buf) // SOCKS version and auth methods
-	if err != nil {
+	buf := make([]byte, 0xff)
```

Comment:
Good point. I thought he double allocation a tradeoff in order to avoid code duplication. There's a way to avoid double allocation and code duplication though: 

1. Write a helper internal function e.g. `readAddr` that takes a buffer argument. 
2. Rewrite the exposed function `ReadAddr` to call `readAddr` with a new buffer.
3. Call `readAddr` in `Handshake` with the existing buffer.


In general we do not change the signature of exposed functions in Go to be backward-compatible.